File: mv-design-pro/backend/src/application/sld/station_geometry.py
"""
Station Geometry — obliczenia geometrii stacji dla SLD.

Rozszerzenie layoutu SLD o:
- StationBoundingBox (NO_ROUTE_RECT): prostokąt stacji na schemacie
- TrunkPath: ścieżka toru głównego SN
- EntryPointMarker: punkt wejścia kabli do stacji

Czyste funkcje: ENM + TopologyGraph → StationGeometry (deterministyczne).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from enm.models import EnergyNetworkModel
from enm.topology import TopologyGraph, TrunkSegment


# ---------------------------------------------------------------------------
# Dataclasses
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class StationBoundingBox:
    """Prostokąt stacji na schemacie SLD (NO_ROUTE_RECT).

    Definiuje strefę, w której trasy kablowe NIE powinny przechodzić.
    """

    substation_ref: str
    station_name: str
    station_type: str
    x: float
    y: float
    width: float
    height: float
    bus_refs: tuple[str, ...]
    bay_count: int


@dataclass(frozen=True)
class TrunkPathSegment:
    """Segment toru głównego SN na schemacie SLD."""

    branch_ref: str
    from_bus_ref: str
    to_bus_ref: str
    order: int
    from_x: float
    from_y: float
    to_x: float
    to_y: float
    length_km: float
    is_highlighted: bool = True


@dataclass(frozen=True)
class EntryPointMarker:
    """Punkt wejścia kabli zewnętrznych do stacji."""

    substation_ref: str
    bus_ref: str
    entry_side: Literal["top", "bottom", "left", "right"]
    x: float
    y: float
    label: str


@dataclass(frozen=True)
class StationGeometry:
    """Pełna geometria stacyjna dla SLD.

    INVARIANT: Ten sam ENM + TopologyGraph → identyczna geometria (determinizm).
    """

    station_boxes: tuple[StationBoundingBox, ...]
    trunk_path: tuple[TrunkPathSegment, ...]
    entry_points: tuple[EntryPointMarker, ...]


# ---------------------------------------------------------------------------
# Builder
# ---------------------------------------------------------------------------


def build_station_geometry(
    enm: EnergyNetworkModel,
    topo: TopologyGraph,
    bus_positions: dict[str, tuple[float, float]],
    *,
    station_padding: float = 40.0,
    station_min_width: float = 200.0,
    station_min_height: float = 160.0,
) -> StationGeometry:
    """
    Zbuduj geometrię stacji z ENM, grafu topologicznego i pozycji szyn.

    Czysta, deterministyczna funkcja.

    Args:
        enm: EnergyNetworkModel — kanoniczny model sieci.
        topo: TopologyGraph — graf topologiczny.
        bus_positions: Mapa ref_id szyny → (x, y) z layoutu SLD.
        station_padding: Margines wokół elementów stacji [px].
        station_min_width: Minimalna szerokość ramki stacji [px].
        station_min_height: Minimalna wysokość ramki stacji [px].

    Returns:
        StationGeometry z pełną informacją geometryczną.
    """
    # 1. Station bounding boxes (NO_ROUTE_RECT)
    station_boxes = _compute_station_boxes(
        enm, bus_positions, station_padding, station_min_width, station_min_height
    )

    # 2. Trunk path segments
    trunk_path = _compute_trunk_path(topo, bus_positions)

    # 3. Entry point markers
    entry_points = _compute_entry_points(enm, bus_positions, station_boxes)

    return StationGeometry(
        station_boxes=tuple(station_boxes),
        trunk_path=tuple(trunk_path),
        entry_points=tuple(entry_points),
    )


# ---------------------------------------------------------------------------
# Internals
# ---------------------------------------------------------------------------


def _compute_station_boxes(
    enm: EnergyNetworkModel,
    bus_positions: dict[str, tuple[float, float]],
    padding: float,
    min_width: float,
    min_height: float,
) -> list[StationBoundingBox]:
    """Oblicz prostokąty stacji (NO_ROUTE_RECT) na podstawie pozycji szyn."""
    boxes: list[StationBoundingBox] = []

    # Zbierz bay count per substation
    bay_counts: dict[str, int] = {}
    for bay in enm.bays:
        bay_counts[bay.substation_ref] = bay_counts.get(bay.substation_ref, 0) + 1

    for sub in sorted(enm.substations, key=lambda s: s.ref_id):
        # Zbierz pozycje szyn w stacji
        bus_coords: list[tuple[float, float]] = []
        for br in sub.bus_refs:
            pos = bus_positions.get(br)
            if pos:
                bus_coords.append(pos)

        if not bus_coords:
            continue

        # Oblicz bounding box
        xs = [c[0] for c in bus_coords]
        ys = [c[1] for c in bus_coords]
        min_x = min(xs) - padding
        max_x = max(xs) + padding
        min_y = min(ys) - padding
        max_y = max(ys) + padding

        width = max(max_x - min_x, min_width)
        height = max(max_y - min_y, min_height)

        # Wycentruj jeśli za mały
        cx = (min(xs) + max(xs)) / 2
        cy = (min(ys) + max(ys)) / 2
        final_x = cx - width / 2
        final_y = cy - height / 2

        boxes.append(StationBoundingBox(
            substation_ref=sub.ref_id,
            station_name=sub.name,
            station_type=sub.station_type,
            x=round(final_x, 1),
            y=round(final_y, 1),
            width=round(width, 1),
            height=round(height, 1),
            bus_refs=tuple(sorted(sub.bus_refs)),
            bay_count=bay_counts.get(sub.ref_id, 0),
        ))

    return boxes


def _compute_trunk_path(
    topo: TopologyGraph,
    bus_positions: dict[str, tuple[float, float]],
) -> list[TrunkPathSegment]:
    """Oblicz ścieżkę toru głównego SN dla wizualizacji."""
    segments: list[TrunkPathSegment] = []

    for trunk in topo.trunk_segments:
        from_pos = bus_positions.get(trunk.from_bus_ref)
        to_pos = bus_positions.get(trunk.to_bus_ref)

        if from_pos and to_pos:
            segments.append(TrunkPathSegment(
                branch_ref=trunk.branch_ref,
                from_bus_ref=trunk.from_bus_ref,
                to_bus_ref=trunk.to_bus_ref,
                order=trunk.order,
                from_x=round(from_pos[0], 1),
                from_y=round(from_pos[1], 1),
                to_x=round(to_pos[0], 1),
                to_y=round(to_pos[1], 1),
                length_km=trunk.length_km,
                is_highlighted=True,
            ))

    return segments


def _compute_entry_points(
    enm: EnergyNetworkModel,
    bus_positions: dict[str, tuple[float, float]],
    station_boxes: list[StationBoundingBox],
) -> list[EntryPointMarker]:
    """Oblicz punkty wejścia kabli do stacji."""
    markers: list[EntryPointMarker] = []
    box_map = {b.substation_ref: b for b in station_boxes}

    for sub in sorted(enm.substations, key=lambda s: s.ref_id):
        if not sub.entry_point_ref:
            continue

        box = box_map.get(sub.ref_id)
        if not box:
            continue

        # Punkt wejścia — domyślnie na górze stacji (strona zasilania)
        entry_x = round(box.x + box.width / 2, 1)
        entry_y = round(box.y, 1)

        markers.append(EntryPointMarker(
            substation_ref=sub.ref_id,
            bus_ref=sub.entry_point_ref,
            entry_side="top",
            x=entry_x,
            y=entry_y,
            label=f"Wejście: {sub.name}",
        ))

    return markers

File: mv-design-pro/frontend/src/engine/sld-layout/station-geometry.ts
/**
 * Station Geometry — typy i obliczenia geometrii stacji dla SLD.
 *
 * Rozszerzenie pipeline SLD o:
 * - StationBoundingBox (NO_ROUTE_RECT): ramka stacji, strefa wyłączona z routingu
 * - TrunkPath: ścieżka toru głównego SN (wyróżniona wizualnie)
 * - EntryPointMarker: punkt wejścia kabli do stacji
 *
 * DETERMINIZM: Ten sam input → identyczny output.
 * BINDING: Etykiety PL, bez nazw kodowych.
 */

import type { Point, Rectangle } from './types';

// =============================================================================
// TYPY — STATION BOUNDING BOX (NO_ROUTE_RECT)
// =============================================================================

/**
 * Prostokąt stacji na schemacie SLD (NO_ROUTE_RECT).
 *
 * Definiuje strefę, w której trasy kablowe NIE powinny przechodzić.
 * Wyrysowywany jako prostokąt z zaokrąglonymi rogami (CSS: border-radius).
 */
export interface StationBoundingBox {
  /** ref_id stacji (Substation) */
  substationRef: string;

  /** Nazwa stacji (do etykiety) */
  stationName: string;

  /** Typ stacji (GPZ, SN/nn, switching, customer) */
  stationType: 'gpz' | 'mv_lv' | 'switching' | 'customer';

  /** Prostokąt ograniczający */
  bounds: Rectangle;

  /** ref_id szyn wewnątrz stacji */
  busRefs: string[];

  /** Liczba pól (bay) w stacji */
  bayCount: number;

  /** Kolor obramowania (zależny od typu stacji) */
  borderColor: string;

  /** Kolor tła (półprzezroczysty) */
  fillColor: string;
}

// =============================================================================
// TYPY — TRUNK PATH (TOR GŁÓWNY)
// =============================================================================

/**
 * Segment toru głównego SN na schemacie SLD.
 *
 * Tor główny = ciąg gałęzi od GPZ do końca magistrali.
 * Wyrysowywany grubszą linią z wyróżnionym kolorem.
 */
export interface TrunkPathSegment {
  /** ref_id gałęzi */
  branchRef: string;

  /** Punkt początkowy */
  from: Point;

  /** Punkt końcowy */
  to: Point;

  /** Kolejność segmentu w torze głównym */
  order: number;

  /** Długość segmentu [km] */
  lengthKm: number;

  /** Czy segment jest wyróżniony (highlight) */
  isHighlighted: boolean;
}

// =============================================================================
// TYPY — ENTRY POINT MARKER
// =============================================================================

/**
 * Punkt wejścia kabli zewnętrznych do stacji.
 *
 * Wizualnie: trójkąt/strzałka na krawędzi ramki stacji.
 */
export interface EntryPointMarker {
  /** ref_id stacji */
  substationRef: string;

  /** ref_id szyny wejściowej */
  busRef: string;

  /** Strona wejścia na ramce stacji */
  entrySide: 'top' | 'bottom' | 'left' | 'right';

  /** Pozycja markera */
  position: Point;

  /** Etykieta (nazwa stacji) */
  label: string;
}

// =============================================================================
// TYPY — PEŁNA GEOMETRIA STACJI
// =============================================================================

/**
 * Pełna geometria stacyjna — overlay na schemacie SLD.
 *
 * INVARIANT: Ten sam ENM → identyczna geometria (determinizm).
 */
export interface StationGeometryResult {
  /** Prostokąty stacji (NO_ROUTE_RECT) */
  stationBoxes: StationBoundingBox[];

  /** Ścieżka toru głównego */
  trunkPath: TrunkPathSegment[];

  /** Punkty wejścia do stacji */
  entryPoints: EntryPointMarker[];
}

// =============================================================================
// KONFIGURACJA KOLORÓW STACJI
// =============================================================================

/**
 * Kolory stacji wg typu — konfigurowalne.
 */
export const STATION_COLORS: Record<
  StationBoundingBox['stationType'],
  { border: string; fill: string }
> = {
  gpz: { border: '#dc2626', fill: 'rgba(220, 38, 38, 0.06)' },
  mv_lv: { border: '#2563eb', fill: 'rgba(37, 99, 235, 0.06)' },
  switching: { border: '#d97706', fill: 'rgba(217, 119, 6, 0.06)' },
  customer: { border: '#059669', fill: 'rgba(5, 150, 105, 0.06)' },
};

/**
 * Konfiguracja toru głównego.
 */
export const TRUNK_PATH_CONFIG = {
  /** Grubość linii toru głównego [px] */
  strokeWidth: 4,
  /** Kolor toru głównego */
  color: '#dc2626',
  /** Styl linii */
  dashArray: 'none',
} as const;

// =============================================================================
// OBLICZENIA (DETERMINISTYCZNE)
// =============================================================================

/**
 * Oblicz geometrię stacji z danych ENM topology.
 *
 * @param substations - Lista substacji z ENM
 * @param bays - Lista pól (bay) z ENM
 * @param busPositions - Mapa ref_id → pozycja (z pipeline layout)
 * @param trunkSegments - Segmenty toru głównego (z TopologyGraph)
 * @param config - Konfiguracja (opcjonalna)
 */
export function computeStationGeometry(
  substations: SubstationInput[],
  bays: BayInput[],
  busPositions: Map<string, Point>,
  trunkSegments: TrunkSegmentInput[],
  config?: Partial<StationGeometryConfig>
): StationGeometryResult {
  const cfg: StationGeometryConfig = { ...DEFAULT_STATION_CONFIG, ...config };

  // 1. Station bounding boxes
  const stationBoxes = computeStationBoxes(substations, bays, busPositions, cfg);

  // 2. Trunk path
  const trunkPath = computeTrunkPath(trunkSegments, busPositions);

  // 3. Entry points
  const entryPoints = computeEntryPoints(substations, busPositions, stationBoxes);

  return { stationBoxes, trunkPath, entryPoints };
}

// =============================================================================
// INPUT TYPES (z ENM)
// =============================================================================

export interface SubstationInput {
  ref_id: string;
  name: string;
  station_type: 'gpz' | 'mv_lv' | 'switching' | 'customer';
  bus_refs: string[];
  entry_point_ref?: string | null;
}

export interface BayInput {
  ref_id: string;
  substation_ref: string;
}

export interface TrunkSegmentInput {
  branch_ref: string;
  from_bus_ref: string;
  to_bus_ref: string;
  order: number;
  length_km: number;
}

export interface StationGeometryConfig {
  /** Margines wokół elementów stacji [px] */
  stationPadding: number;
  /** Minimalna szerokość ramki stacji [px] */
  stationMinWidth: number;
  /** Minimalna wysokość ramki stacji [px] */
  stationMinHeight: number;
}

const DEFAULT_STATION_CONFIG: StationGeometryConfig = {
  stationPadding: 40,
  stationMinWidth: 200,
  stationMinHeight: 160,
};

// =============================================================================
// INTERNALS
// =============================================================================

function computeStationBoxes(
  substations: SubstationInput[],
  bays: BayInput[],
  busPositions: Map<string, Point>,
  config: StationGeometryConfig
): StationBoundingBox[] {
  const boxes: StationBoundingBox[] = [];

  // Bay count per substation
  const bayCounts = new Map<string, number>();
  for (const bay of bays) {
    bayCounts.set(bay.substation_ref, (bayCounts.get(bay.substation_ref) ?? 0) + 1);
  }

  // Sort for determinism
  const sortedSubs = [...substations].sort((a, b) => a.ref_id.localeCompare(b.ref_id));

  for (const sub of sortedSubs) {
    const coords: Point[] = [];
    for (const busRef of sub.bus_refs) {
      const pos = busPositions.get(busRef);
      if (pos) coords.push(pos);
    }
    if (coords.length === 0) continue;

    const xs = coords.map((c) => c.x);
    const ys = coords.map((c) => c.y);

    const rawWidth = Math.max(...xs) - Math.min(...xs) + config.stationPadding * 2;
    const rawHeight = Math.max(...ys) - Math.min(...ys) + config.stationPadding * 2;
    const width = Math.max(rawWidth, config.stationMinWidth);
    const height = Math.max(rawHeight, config.stationMinHeight);

    const cx = (Math.min(...xs) + Math.max(...xs)) / 2;
    const cy = (Math.min(...ys) + Math.max(...ys)) / 2;

    const colors = STATION_COLORS[sub.station_type];

    boxes.push({
      substationRef: sub.ref_id,
      stationName: sub.name,
      stationType: sub.station_type,
      bounds: {
        x: Math.round((cx - width / 2) * 10) / 10,
        y: Math.round((cy - height / 2) * 10) / 10,
        width: Math.round(width * 10) / 10,
        height: Math.round(height * 10) / 10,
      },
      busRefs: [...sub.bus_refs].sort(),
      bayCount: bayCounts.get(sub.ref_id) ?? 0,
      borderColor: colors.border,
      fillColor: colors.fill,
    });
  }

  return boxes;
}

function computeTrunkPath(
  trunkSegments: TrunkSegmentInput[],
  busPositions: Map<string, Point>
): TrunkPathSegment[] {
  const segments: TrunkPathSegment[] = [];

  // Sort by order for determinism
  const sorted = [...trunkSegments].sort((a, b) => a.order - b.order);

  for (const trunk of sorted) {
    const fromPos = busPositions.get(trunk.from_bus_ref);
    const toPos = busPositions.get(trunk.to_bus_ref);

    if (fromPos && toPos) {
      segments.push({
        branchRef: trunk.branch_ref,
        from: { x: Math.round(fromPos.x * 10) / 10, y: Math.round(fromPos.y * 10) / 10 },
        to: { x: Math.round(toPos.x * 10) / 10, y: Math.round(toPos.y * 10) / 10 },
        order: trunk.order,
        lengthKm: trunk.length_km,
        isHighlighted: true,
      });
    }
  }

  return segments;
}

function computeEntryPoints(
  substations: SubstationInput[],
  _busPositions: Map<string, Point>,
  stationBoxes: StationBoundingBox[]
): EntryPointMarker[] {
  const markers: EntryPointMarker[] = [];
  const boxMap = new Map(stationBoxes.map((b) => [b.substationRef, b]));

  const sorted = [...substations].sort((a, b) => a.ref_id.localeCompare(b.ref_id));

  for (const sub of sorted) {
    if (!sub.entry_point_ref) continue;

    const box = boxMap.get(sub.ref_id);
    if (!box) continue;

    // Default entry point: top center of station box
    markers.push({
      substationRef: sub.ref_id,
      busRef: sub.entry_point_ref,
      entrySide: 'top',
      position: {
        x: Math.round((box.bounds.x + box.bounds.width / 2) * 10) / 10,
        y: Math.round(box.bounds.y * 10) / 10,
      },
      label: `Wejście: ${sub.name}`,
    });
  }

  return markers;
}
